docs/src/developing/backwards-compatibility.md

```diff
@@ -155,3 +165,81 @@ circumvented in order to rapidly deploy a fix, depending on the severity of the
 CLI tooling json output (`output --json`) compatibility will be preserved; however, output directed
 for a human reader is subject to change. This includes output as well as potential help, warning, or
 error messages.
+
+### Why not just use SemVer
+
+The Solana Rust SDK crates do *not* follow semantic versioning (SemVer) for
+breaking changes.
+
+#### SemVer in Rust
+
+Under SemVer, breaking changes, such as removal of functions or types, only happens
+in a new major version.
+
+In many situations, this is useful -- it is undesirable to pick up a breaking
+change accidentally. The default dependency declaration in Cargo follows SemVer.
+It assumes that all minor versions are compatible, and will automatically update
+to the newest minor version available.
+
+For example, if you declare `solana-program = "1.10"`, Cargo can pull in version
+`1.16`, since it assumes that all `1.X.Y` releases are compatible.
+
+If, however, Cargo deems that two versions of a package are incompatible, it will
+treat them as two completely separate packages.
+
+Packages can be deemed incompatible for many reasons, but the most common situation
+under the default declaration format (SemVer) is due to different major versions.
+
+If a package ends up with two or more instances of `solana-program`,
+[bad things will happen](https://doc.rust-lang.org/cargo/reference/resolver.html#version-incompatibility-hazards),
+either at compile-time or runtime, if the conflicting types in the public APIs
+are ever used together.
+
+Because of these issues, developers should only want one instance of `solana-program`
+in their project. This includes *all* direct and indirect dependencies: they should
+all use the same `solana-program`.
+
+If developers all use the default versioning declaration with major version `1`,
+then we can ensure that only one version of `solana-program` exists in the resolution.
+
+The specifics of the Cargo resolver are beyond the scope of this document.
+You can find more information about the Cargo resolver in
+[The Cargo Book](https://doc.rust-lang.org/cargo/reference/resolver.html).
+
+#### So what's your solution?
+
+We would like all releases to be 100% backwards compatible, allowing everyone
+to upgrade at any time with no effort.  This model, however, would put overly
+strict constraints on the development of the Rust SDK crates, considering we cannot
+in practice increase the major version.
```

Comment:
I'm not sure I understand. All the reasons for a different method than semver here seems to lead to that you `cannot in practice increase the major version`

Why is that?

Reply:
This was covered in a previous commit, but I removed it. Adding it here for more motivation:

> Cargo will force all crates in a build to use the same major version
of a dependency. For example, if your crate `my-crate` declares `solana-program = "1.10"`,
and a dependency `my-dependency` declares `solana-program = "1.16"`, both your crate and your
dependency will be built with version `1.16`, at least.

> Additionally, the types in `solana-program` used by both your crate and your
dependency are treated as the same.

> For example, if `solana-program` has a struct declared as:

```
#[derive(PartialEq)]
pub struct Pubkey(pub [u8; 32]);
```

> And `my-dependency` has:

```
pub const DEPENDENCY_PUBKEY = Pubkey([1; 32]);
```

> Then `my-crate` can say:

```
use solana_program::Pubkey;
let my_pubkey = Pubkey([1; 32]);
if my_pubkey == my_dependency::DEPENDENCY_PUBKEY {
    println!("They're the same");
}
```

> On the other hand, types from crates with different major versions are treated as
completely different types. If `my-crate` declares `solana-program = "2"` and
`my-dependency` declares `solana-program = "1"`, then the preceding code *will not*
compile.

> To get this code to compile, you need to create functions to convert between the
types in v1 and v2. And whenever v3 is released, the entire problem becomes multiplied.

But either way, I don't think we should move forward with changing this. I'll provide an update in the main semver issue.